backfill/backfill.go

```diff
@@ -0,0 +1,399 @@
+	ctx, span := tracer.Start(ctx, "BackfillRepo")
+	defer span.End()
+
+	start := time.Now()
+
+	repoDid := job.Repo()
+
+	log := slog.With("source", "backfiller_backfill_repo", "repo", repoDid)
+	log.Info(fmt.Sprintf("processing backfill for %s", repoDid))
+
+	var url = fmt.Sprintf("%s?did=%s", b.CheckoutPath, repoDid)
+
+	// GET and CAR decode the body
+	client := &http.Client{
+		Transport: otelhttp.NewTransport(http.DefaultTransport),
+		Timeout:   120 * time.Second,
+	}
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		log.Error("failed to create request", "error", err)
+		return
+	}
+
+	req.Header.Set("Accept", "application/vnd.ipld.car")
+	req.Header.Set("User-Agent", fmt.Sprintf("atproto-backfill-%s/0.0.1", b.Name))
+	if b.magicHeaderKey != "" && b.magicHeaderVal != "" {
+		req.Header.Set(b.magicHeaderKey, b.magicHeaderVal)
+	}
+
+	b.syncLimiter.Wait(ctx)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Error("failed to send request", "error", err)
+		return
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		log.Info("failed to get repo", "status", resp.StatusCode)
+		reason := "unknown error"
+		if resp.StatusCode == http.StatusBadRequest {
+			reason = "repo not found"
+		}
+		state := fmt.Sprintf("failed (%s)", reason)
+
+		// Mark the job as "failed"
+		err := job.SetState(ctx, state)
+		if err != nil {
+			log.Error("failed to set job state", "error", err)
+		}
+
+		// Clear buffered ops
+		err = job.ClearBufferedOps(ctx)
+		if err != nil {
+			log.Error("failed to clear buffered ops", "error", err)
+		}
+		return
+	}
+
+	instrumentedReader := instrumentedReader{
+		source:  resp.Body,
+		counter: backfillBytesProcessed.WithLabelValues(b.Name),
+	}
+
+	defer instrumentedReader.Close()
+
+	r, err := repo.ReadRepoFromCar(ctx, instrumentedReader)
+	if err != nil {
+		log.Error("failed to read repo from car", "error", err)
+
+		state := "failed (couldn't read repo CAR from response body)"
+
+		// Mark the job as "failed"
+		err := job.SetState(ctx, state)
+		if err != nil {
+			log.Error("failed to set job state", "error", err)
+		}
+
+		// Clear buffered ops
+		err = job.ClearBufferedOps(ctx)
+		if err != nil {
+			log.Error("failed to clear buffered ops", "error", err)
+		}
+		return
+	}
+
+	numRecords := 0
+	numRoutines := b.ParallelRecordCreates
+	recordQueue := make(chan recordQueueItem, numRoutines)
+	recordResults := make(chan recordResult, numRoutines)
+
+	wg := sync.WaitGroup{}
+
+	// Producer routine
+	go func() {
+		defer close(recordQueue)
+		r.ForEach(ctx, b.NSIDFilter, func(recordPath string, nodeCid cid.Cid) error {
+			numRecords++
+			recordQueue <- recordQueueItem{recordPath: recordPath, nodeCid: nodeCid}
+			return nil
+		})
+	}()
+
+	// Consumer routines
+	for i := 0; i < numRoutines; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for item := range recordQueue {
+				recordCid, rec, err := r.GetRecord(ctx, item.recordPath)
```

Comment:
TIL: Repo is threadsafe

Reply:
you could also just load the record by CID directly instead of having to do a path traversal

Reply:
I changed to invoke the blockstore directly and GET the CID but I have to do a little typecast that's kinda annoying, should be cheaper though